Update call to amp.autocast from fast_dtype to dtype

jstjohn opened the pull request:
## What does this PR do?

In a recent commit in pytorch 1.10, [the signature for `autocast` changed from `autocast(fast_dtype=...)` to `autocast(dtype=...)`.](https://github.com/pytorch/pytorch/commit/6324d98e9e736214ee7a161482b02aaf8cebbc9d#diff-dac4bd53ced015c8810b3b02fc5c2ec6c2b0658c5090b4fbbd09c96bd45087d1R16-R17) This PR reflects that signature change in pytorch lightning, since 1.10 is needed for `bfloat16` types.

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Fixes #9210

### Does your PR introduce any breaking changes? If yes, please list them.

This breaks backwards compatibility with any pytorch version earlier than 1.10. It is needed if they stick with this name change in 1.10 and we want to support `bfloat16` types though.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes? (I ran `pytest tests/models/test_amp.py -v`)
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Yup!

carmocca commented:
I think it's worth waiting on this until https://github.com/pytorch/pytorch/pull/64122 is merged to torch master

jstjohn commented on tests/models/test_amp.py:
```diff
@@ -108,7 +108,7 @@ def test_amp_cpus(tmpdir, accelerator, precision, num_processes):
     [
         None,
-        pytest.param("dp", marks=pytest.mark.skip("dp + amp not supported currently")),  # TODO
+        "dp",
```
On my system (two A10), this test passed for me (both 1 and 2 GPUs) with bfloat16.

jstjohn commented on tests/models/test_amp.py:
```diff
@@ -74,7 +74,7 @@ def _assert_autocast_enabled(self):
     [
         None,
-        pytest.param("dp", marks=pytest.mark.skip("dp + amp not supported currently")),  # TODO
+        pytest.param("dp", marks=pytest.mark.skip("dp + amp not supported on CPU currently")),  # TODO
```
This test still fails for me, but since it only seems to happen on the CPU I thought it would be a good idea to call that out.

tchaton replied:
Yes, this need some investigation. Thanks !

carmocca commented on pytorch_lightning/plugins/precision/native_amp.py:
```diff
@@ -109,9 +109,9 @@ def pre_optimizer_step(
     def autocast_context_manager(self) -> torch.cuda.amp.autocast:
         if self.use_cpu:
-            return torch.cpu.amp.autocast(fast_dtype=self._fast_dtype)
+            return torch.cpu.amp.autocast(dtype=self._fast_dtype)
```
Can you also update the `self` attribute?

```python
pytorch_lightning/plugins/precision/native_amp.py:55:        self._fast_dtype = self._select_precision_dtype(precision)
pytorch_lightning/plugins/precision/native_amp.py:75:        return self._fast_dtype == torch.bfloat16
pytorch_lightning/plugins/precision/native_amp.py:112:            return torch.cpu.amp.autocast(fast_dtype=self._fast_dtype)
pytorch_lightning/plugins/precision/native_amp.py:114:            return torch.cuda.amp.autocast(fast_dtype=self._fast_dtype)
tests/plugins/test_amp_plugins.py:220:    assert context_manager.fast_dtype == torch.bfloat16
```

tchaton replied:
Should we use the PyTorch version and use fast_dtype or dtype accordingly ? Would this work with work with lower version of PyTorch ?

jstjohn replied:
This will almost surely start breaking lower versions of pytorch. Then again bfloat16 will not work there... I guess this would break amp.autocast for `use_cpu` in general though for older versions of pytorch. Annoying.

jstjohn replied:
Actually when use_cpu is set, there is another check for pytorch 1.10 or higher, so that line would not fail for earlier versions since neither of those two amp.autocast lines should not be called unless either:

1) self.use_cpu is set
2) self.is_bfloat16 is set

use_cpu is enforced on pytorch 1.10+ in init:

```
        if use_cpu and not _TORCH_CPU_AMP_AVAILABLE:
            raise MisconfigurationException(
                "You have asked for native AMP on CPU, but AMP is only available on GPU for PyTorch 1.9 "
                "and lower. To use native AMP on CPU, install PyTorch 1.10 or later."
            )
```

and self.is_bfloat16 should only happen for pytorch 1.10+ as well. 

For other cases, it looks like the context manager doesn't have dtype (or previously fast_dtype) set at all, so it _should_ work.

So if this does create problems, it would be due to a bug and not a known issue.

jstjohn commented:
> I think it's worth waiting on this until [pytorch/pytorch#64122](https://github.com/pytorch/pytorch/pull/64122) is merged to torch master

@carmocca Looks like they just merged that PR you found into master.

jstjohn commented:
> > Looks like they just merged that PR you found into master.
> 
> Cool! Can you update this PR to require that fix?

Hmm... Wonder how long it is going to take for the merged PR to land in nightly. It's not in there as of today:
```python
>>> torch.__version__
'1.10.0.dev20210901'
>>> torch.cuda.is_bf16_supported()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jstjohn/miniconda3/envs/pl_dev/lib/python3.9/site-packages/torch/cuda/__init__.py", line 85, in is_bf16_supported
    cuda_maj_decide = int(cu_vers.split(',')[0]) >= 11
ValueError: invalid literal for int() with base 10: '11.1'
```

jstjohn commented:
@carmocca it is fixed now, I just updated to the 0902 dev nightly and the fix is merged into pytorch. I updated the version requirement arg as well. `pytest tests/models/test_amp.py -v` is all passing (with the same skips as described before, minus the gpu + dp/ddp + bfloat16 tests which now work).

```python
>>> import torch
>>> torch.__version__
'1.10.0.dev20210902'
>>> torch.cuda.is_bf16_supported()
True
```

tchaton reviewed (approved):
LGTM ! Still failing conda test on Torch 1.10

jstjohn commented:
@tchaton hey sorry about this, I think a comment string I pushed to disable the DeepSource error with the argument names switching in 1.10 also removed the auto-merge flag you set.

carmocca commented on tests/models/test_amp.py:
```diff
@@ -108,7 +108,7 @@ def test_amp_cpus(tmpdir, accelerator, precision, num_processes):
         None,
-        pytest.param("dp", marks=pytest.mark.skip("dp + amp not supported currently")),  # TODO
+        "dp",
         "ddp_spawn",
```
```suggestion
        "ddp_spawn"
```
